lib/mixin/vpp.py

```diff
@@ -160,18 +160,21 @@ def check_stack(self):
     params = vars(self).copy()
 
     if self.stack in ["xstack"]:
-      params.update(metric = dict(type = "ssim", miny = 0.97, minu = 0.97, minv = 0.97))
+      params.setdefault("metric", dict(type = "ssim", miny = 0.97, minu = 0.97, minv = 0.97))
       params.update(width = self.cols * self.tilew, height = self.rows * self.tileh)
-      params.update(reference = format_value(self.reference, **params))
     elif self.stack in ["hstack"]:
-      params.update(metric = dict(type = "md5"))
+      params.setdefault("metric", dict(type = "md5"))
       params.update(width = self.inputs * self.width, height = self.height)
     elif self.stack in ["vstack"]:
-      params.update(metric = dict(type = "md5"))
+      params.setdefault("metric", dict(type = "md5"))
       params.update(width = self.width, height = self.height * self.inputs)
     else:
       assert False, f"unknown stack operation: '{self.stack}'"
 
+    # a reference is needed by some metrics
+    if vars(self).get("reference", None) is not None:
+      params.update(reference = format_value(self.reference, **params))
+
```

Comment:
Hmm, I guess this won't work with our test configs b/c the reference template assumes xstack :/  Hence the check failures.  We need to think of another way to handle this.  I will try to take look into it next week.

Reply:
Maybe we can use the post validate in test config to set the reference based on stack mode.

Reply:
Ok, I see you've updated the test config

Reply:
Please send patches for bat and full test configs.

Reply:
ok , will update these part, thanks for your help and review.